R/credentials_gce.R

```diff
@@ -74,36 +74,60 @@
 #' }
 credentials_gce <- function(scopes = "https://www.googleapis.com/auth/cloud-platform",
                             service_account = "default", ...) {
-  gargle_debug("trying {.fun credentials_gce}")
+  gargle_debug("Trying {.fun credentials_gce} ...")
   if (!is_gce()) {
+    gargle_debug(c("x" = "We don't seem to be on GCE."))
     return(NULL)
   }
 
-  gce_token <- fetch_gce_access_token(scopes, service_account = service_account)
+  requested_scopes <- normalize_scopes(scopes)
+  dat <- gce_instance_service_accounts()
+  service_account_details <- as.list(dat[dat$name == service_account, ])
+
+  account_scopes <- service_account_details$scopes
+  account_scopes <- normalize_scopes(strsplit(account_scopes, split = ",")[[1]])
+  missing <- setdiff(requested_scopes, account_scopes)
+  if (length(missing) > 0) {
+    gargle_debug(c(
+      "!" = "{cli::qty(length(missing))}{?This/These} requested \\
+             scope{?s} {?is/are} not among the scopes for the \\
+             {.val {service_account}} service account:",
+      bulletize(missing, bullet = "x"),
+      "i" = "If there are problems downstream, this might be the root cause."
+    ))
+  }
+
+  token <- gce_access_token(scopes, service_account = service_account)
 
-  params <- list(
-    as_header = TRUE,
-    scope = scopes,
-    service_account = service_account
-  )
-  token <- GceToken$new(
-    credentials = gce_token$access_token,
-    params = params,
-    cache_path = FALSE,
-    # The underlying Token2 class appears to *require* an endpoint and an app,
-    # though it doesn't use them for anything in this case.
-    endpoint = httr::oauth_endpoints("google"),
-    app = httr::oauth_app("google", key = "KEY", secret = "SECRET")
-  )
-  token$refresh()
```

Comment:
This logic has been relocated as I refactor this to look more like all the other token flows.

Various simplifications are also happening at the same time:

* I no longer need to set `cache_path = FALSE` or to to set the `endpoint` and `app` in light of the broader changes to the `GceToken` class.
* I don't call `token$refresh()` here anymore. Let's talk about that!

The original implementation basically fetched a token (token one) then immediately refreshed that token, which redoes the whole flow (token two). I _think_ that is not necessary? But would love to hear if I am wrong.

I think, if the current access token is found to have expired, downstream requests will see that this token is refreshable and will refresh (i.e. start over) as needed.